spring-cloud-stream/src/main/java/org/springframework/cloud/stream/function/BindableFunctionProxyFactory.java

```diff
@@ -55,25 +65,73 @@ public void afterPropertiesSet() {
 		Assert.notEmpty(BindableFunctionProxyFactory.this.bindingTargetFactories,
 				"'bindingTargetFactories' cannot be empty");
 
+		this.multiple = this.inputCount > 1 || this.outputCount > 1;
+
 		if (this.inputCount > 0) {
-			if (this.inputCount == 1) {
-				this.createInput("input");
+			if (multiple || nameBasedOnFunctionName) {
+				for (int i = 0; i < inputCount; i++) {
+					this.createInput(this.buildInputNameForIndex(i));
+				}
 			}
 			else {
-				throw new UnsupportedOperationException("Multiple inputs are not currently supported");
+				this.createInput("input");
 			}
 		}
 
 		if (this.outputCount > 0) {
-			if (this.outputCount == 1) {
-				this.createOutput("output");
+			if (multiple || nameBasedOnFunctionName) {
+				for (int i = 0; i < outputCount; i++) {
+					this.createOutput(this.buildOutputNameForIndex(i));
+				}
 			}
 			else {
-				throw new UnsupportedOperationException("Multiple outputs are not currently supported");
+				this.createOutput("output");
 			}
 		}
 	}
 
+	@Override
+	public Class<?> getObjectType() {
+		return this.type;
+	}
+
+	@Override
+	public boolean isSingleton() {
+		return true;
+	}
+
+	protected boolean isNameBasedOnFunctionName() {
+		return this.nameBasedOnFunctionName;
+	}
+
+	protected String getFunctionDefinition() {
+		return this.functionDefinition;
+	}
+
+	protected String getInputName(int index) {
+		return CollectionUtils.isEmpty(this.getInputs())
+				? null
+						: this.getInputs().toArray(new String[] {})[index];
```

Comment:
Looks like ` buildInputNameForIndex(int index)` produces exactly the same, but doesn't create an extra array in between.
Anyway it is recommended to use `new String[0] for `toArray()`...

Reply:
True, but `getInputName` returns the cached version, while he other one builds a new string so it could be added to `input/outputHolders` (existing functionality from BindableProxyFactory)